CAM-10801: filtering by candidateUser in an OR query

palossyl opened the pull request:
The [TaskQueryImpl.java](https://github.com/camunda/camunda-bpm-platform/blob/7.12.0-alpha3/engine/src/main/java/org/camunda/bpm/engine/impl/TaskQueryImpl.java) class changed, and the [TaskCandidateTest.java](https://github.com/camunda/camunda-bpm-platform/blob/7.12.0-alpha3/engine/src/test/java/org/camunda/bpm/engine/test/bpmn/usertask/TaskCandidateTest.java) class got a new testcase.

* the `getCandidateGroups()` method now resolves the `candidateUser`'s groups in case of an 'or' query too, so the TaskQuery returns correct results.
* the `getCandidateGroups()` is working for non 'or' queries exactly the same way as before.
* the `getGroupsForCandidateUser(...)` method caches it's results now, as suggested in the implementation notes.

Solves [CAM-10801](https://app.camunda.com/jira/browse/CAM-10801).

palossyl commented:
The `TaskCandidateTest.java` class has got a new testcase to test the conditions where the the error occured before.

tasso94 commented:
Hi palossyl,

thank you for raising this pull request. We will look into it and come back to you as soon as possible.

Stay tuned!

Cheers,
Tassilo

palossyl commented:
Hi @tasso94 !

Thank you! As you have time for it!

Cheers:
          László

tasso94 commented:
Hi @palossyl,

I reviewed your contribution:
* Design and code wise it looks good
* Could you please format your contribution according to our [Code Style Guideline](https://github.com/camunda/camunda-bpm-dev-docs/blob/master/development/Coding-Style-Java.md)?

Cheers,
Tassilo

palossyl commented:
Hi @tasso94!

I have found the code formatter file for eclipse here: [formatter.xml](https://github.com/camunda/camunda-bpm-platform/blob/master/settings/eclipse/formatter.xml). 
The new and changed sections in the code of the two files (`TaskQueryImpl.java`, and `TaskCandidateTest.java`) were formatted with that formatter.
Other sections of the files were left unchanged for you to see only the relevant changes with a compare tool.
There is a second commit now, but only a comment (with little meaning) was deleted from `TaskQueryImpl.java`.

If you like it, i can format the two whole files with the formatter?

`TaskQueryImpl.java` will have many (~80) cosmetic changes, like tabs instead of spaces, much 'empty' lines contain a tab character, bracketing of `else if` statments are not according the style guide, and other minor formatting fixes like that!

Remarks:

The `setting org.eclipse.jdt.core.formatter.join_wrapped_lines` of `formatter.xml` shoud be set to `false`!
To not to join lines in sections sections like:
```java
  List<Task> tasks = taskService
    .createTaskQuery()
    .taskAssignee(KERMIT)
    .list();
```
I have seen tahat an other repository uses [.editorconfig](https://github.com/camunda/camunda-bpm-spring-boot-starter/blob/master/.editorconfig), perhaps for the non eclipse users.

Wich is your preferred way to format the code automatically?

To be short :), the new and changed code as it is committed now shoud agree with the code style guideline!

Cheers:
             László

tasso94 commented:
Hi László,

I'm afraid that the mentioned formatter files are outdated and not maintained anymore.

I would like to propose that you format only the lines that you have touched according to our [Code Style Guideline](https://github.com/camunda/camunda-bpm-dev-docs/blob/master/development/Coding-Style-Java.md): 
* For instance, the language level you are using is Java 6; Please make sure to use the language level constructs of Java 7
* Also `private` methods are not allowed; Use `protected` instead

Cheers,
Tassilo

palossyl commented:
Hi Tassilo!

Thank you for the hints, changed to empty diamons for 'new' on the right side, and no private method.
Please check it again!

Cheers:
      László

tasso94 commented:
Hi László,

I've just merged your contribution.

Thank you!

Cheers,
Tassilo

palossyl commented:
Hi Tassilo!

Thank you for your support!

Cheers:
              László
